Port CSharpSymbolTests to run on Desktop and CoreCLR.

agocke opened the pull request:
This change retargets the CSharpSymbolTest project from .NET 4.6 (which runs only on the desktop CLR) to netstandard 1.3 (which runs on both the Desktop 4.6 framework and CoreCLR).

These changes are necessary to enable more test coverage on CoreCLR, which will be especially impactful when we want to test language features which are only targeted to run on CoreCLR.

There are a lot of file changes in this PR but the most significant new code is in the TestExecutionContext, which is a new CoreCLR AppContext-based mechanism for running in-memory assemblies. This class is meant to have similar functionality to the existing desktop RuntimeEnvironment, but because AppDomains and AppContexts are incompatible, there are some substantial differences. Most notably, the TestExecutionContext views all assemblies next to the CoreCLR test runner (the "app base") as being shared among all test execution runs, including static variables. In contrast, the AppDomain runner only shares assemblies which are already loaded.

The second notable change is a rename from `CreateCompilationWithMscorlib` to `CreateStandardCompilation`. Despite the name, the most common usage of `CreateCompilationWithMscorlib` was just meant to create a simple compilation. The .NET Framework mscorlib was provided as a reference because it provided most of the common types our tests need to compile against, but it wasn't particularly important which types it supplied, beyond a minimum set.

Unfortunately, compiling against the .NET Framework mscorlib can introduce severe incompatibilities when trying to run the application on CoreCLR (which uses System.Runtime as its basis). This PR instead provides a new set of references (NETStandard.Library 2.0, the "standard" references) to try to provide as much compatibility as possible while also providing portability. Thus, the `CreateCompilationWithMscorlib` methods were renamed to `CreateStandardCompilation` to make it clear that no specific set of references is guaranteed when compiling against it. If that is desired, it's recommended to use the existing `CreateCompilation` methods to construct an explicit set of references to compile against (with all the implications that come from that).

agocke commented:
test mac please

agocke commented:
test mac please

agocke commented:
test mac please

agocke commented:
cc @jaredpar @dotnet/roslyn-compiler @tmat For review

I'd suggest reviewing commit by commit and skipping the second commit (mechanical rename).

jaredpar commented on build/Targets/Settings.props:
```diff
@@ -31,6 +31,12 @@
+    <NetStandard20RefPath>$(NuGetPackageRoot)/NETStandard.Library/$(NetStandard20LibraryVersion)/build/netstandard2.0/ref/</NetStandard20RefPath>
+
+    <MicrosoftCSharpVersion>4.3.0</MicrosoftCSharpVersion>
+    <MicrosoftCSharpRefPath>$(NuGetPackageRoot)/Microsoft.CSharp/$(MicrosoftCSharpVersion)/ref/netstandard1.0/</MicrosoftCSharpRefPath>
```
Why are we specifying thes properties here?  Shouldn't they be generated into Dependencies.props?

jaredpar commented on build/Targets/Imports.targets:
```diff
@@ -453,6 +453,13 @@
+             BuildInParallel="true">
+
+        <Output TaskParameter="TargetOutputs" ItemName="ReferenceCopyLocalPaths" />
+    </MSBuild>
```
I was wondering how you would work around this problem.  Can you add a comment documenting what this accomplishes?

jaredpar commented on Compilers.sln:
```diff
@@ -206,10 +206,6 @@ Global
-		{F7712928-1175-47B3-8819-EE086753DEE2}.Debug|Any CPU.ActiveCfg = Debug|Any CPU
-		{F7712928-1175-47B3-8819-EE086753DEE2}.Debug|Any CPU.Build.0 = Debug|Any CPU
-		{F7712928-1175-47B3-8819-EE086753DEE2}.Release|Any CPU.ActiveCfg = Release|Any CPU
-		{F7712928-1175-47B3-8819-EE086753DEE2}.Release|Any CPU.Build.0 = Release|Any CPU
```
Do you know why this changed?

jaredpar commented:
>             s_msbuildDirectory = Roslyn.Test.Utilities.Desktop.TestHelpers.GetMSBuildDirectory();

Seems like we've switched to a lot of fully qualified name invocations here.  Why didn't we add say DesktopTestHelpers to avoid this qualification? 

---
Refers to: src/Compilers/Core/MSBuildTaskTests/IntegrationTests.cs:21 in da5ba75. [](commit_id = da5ba752b4870d77fd8495e79f15c09aa3873838, deletion_comment = False)

jaredpar commented:
>             : new EmitOptions().WithDebugInformationFormat(DebugInformationFormat.PortablePdb);

I thought this was going to go into master as a separate PR so we could discuss the compat issues.  Should be a quick discussion though. 

---
Refers to: src/Compilers/Core/Portable/Emit/EmitOptions.cs:17 in da5ba75. [](commit_id = da5ba752b4870d77fd8495e79f15c09aa3873838, deletion_comment = False)

jaredpar commented on src/Compilers/CSharp/Test/Emit/BreakingChanges.cs:
```diff
@@ -35,7 +35,7 @@ public void Do()
 }
 ";
-            var comp = CreateCompilationWithMscorlib(text);
+            var comp = CreateStandardCompilation(text);
```
Definitely going to have to send an email out about this.  Makes complete sense but need to make sure that going forward we get people to begin to use this is the deafult.

jaredpar commented:
>         public static bool IsRunningOnCoreClr => AssemblyLoadContext.Type != null;

Other members are internal here, why is this public? 

---
Refers to: src/Compilers/Shared/CoreClrShim.cs:14 in da5ba75. [](commit_id = da5ba752b4870d77fd8495e79f15c09aa3873838, deletion_comment = False)

jaredpar commented:
> using DesktopTestHelpers = Roslyn.Test.Utilities.Desktop.TestHelpers;

Looks like DesktopTestHelpers would be a good name for this type :smile:

---
Refers to: src/EditorFeatures/TestUtilities/TestExportProvider.cs:12 in da5ba75. [](commit_id = da5ba752b4870d77fd8495e79f15c09aa3873838, deletion_comment = False)

jaredpar commented:
>         private static ImmutableArray<MetadataReference> s_mscorlibRefArray = ImmutableArray.Create(MscorlibRef);

readonly?

---
Refers to: src/Compilers/Test/Utilities/CSharp/CSharpTestBase.cs:528 in da5ba75. [](commit_id = da5ba752b4870d77fd8495e79f15c09aa3873838, deletion_comment = False)

jaredpar commented on src/Test/Utilities/Portable/FX/CappedStringWriter.cs:
```diff
@@ -0,0 +1,78 @@
+            }
+        }
+
+        public int Length => GetStringBuilder().Length;
```
Prefer properties at top of class after fields

jaredpar commented on src/Test/Utilities/Portable/Compilation/IRuntimeEnvironment.cs:
```diff
@@ -28,28 +29,26 @@ private static IRuntimeEnvironmentFactory GetFactoryImplementation()
+            {
+                runtimeEnvironment.CaptureOutput(action, expectedLength, out output, out errorOutput);
+            }
         }
```
Consider using tuple return vs. double out parameter.

jaredpar commented on src/Test/Utilities/Portable/Compilation/IRuntimeEnvironment.cs:
```diff
@@ -67,6 +66,20 @@ internal EmitOutput(ImmutableArray<byte> assembly, ImmutableArray<byte> pdb)
 
     internal static class RuntimeUtilities
     {
+        internal static T GetFactoryImplementation<T>(string assemblyName, string typeName)
```
Methods should come after fields.

jaredpar commented on src/Test/Utilities/Portable/ConditionalFactAttribute.cs:
```diff
@@ -1,32 +1,9 @@
-using Xunit;
 using Roslyn.Utilities;
 
 namespace Roslyn.Test.Utilities
```
This file looks empty now.  Can we delete it?

jaredpar commented:
> using System;

Missing copyright

---
Refers to: src/Test/Utilities/CoreClr/CoreCLRRuntimeEnvironmentFactory.cs:1 in 37456ea. [](commit_id = 37456ead53c255ab3c8feb6657b3c7e645428d98, deletion_comment = False)

jaredpar commented on src/Test/Utilities/CoreClr/CoreCLRRuntimeEnvironment.cs:
```diff
@@ -68,12 +109,15 @@ public SortedSet<string> GetMemberSignaturesFromMetadata(string fullyQualifiedTy
         public string[] PeVerifyModules(string[] modulesToVerify, bool throwOnError = true)
         {
-            throw new NotImplementedException();
+            // TODO: Implement peverify
```
Bug?

jaredpar commented on src/Test/Utilities/CoreClr/TestExecutionLoadContext.cs:
```diff
@@ -0,0 +1,184 @@
+    internal sealed class TestExecutionLoadContext : AssemblyLoadContext
+    {
+        private readonly static ImmutableDictionary<string, string> s_platformAssemblyPaths = GetPlatformAssemblyPaths();
+        private readonly static Dictionary<string, Assembly> s_loadedPlatformAssemblies = new Dictionary<string, Assembly>();
```
>(); [](start = 122, length = 3)

Consider using an explicit string comparer here.

jaredpar commented on src/Test/Utilities/CoreClr/TestExecutionLoadContext.cs:
```diff
@@ -0,0 +1,184 @@
+
+        public TestExecutionLoadContext(IList<ModuleData> dependencies)
+        {
+            _dependencies = new Dictionary<string, ModuleData>(dependencies.Count);
```
>dependencies.Count [](start = 63, length = 18)

Consider using an explicit string comparer here.

jaredpar commented on src/Test/Utilities/CoreClr/TestExecutionLoadContext.cs:
```diff
@@ -0,0 +1,184 @@
+            var mainAssembly = LoadImageAsAssembly(mainImage);
+            var entryPoint = mainAssembly.EntryPoint;
+
+            Debug.Assert(entryPoint != null, "Attempting to execute an assembly that has no entrypoint; is your test trying to execute a DLL?");
```
Why not make this a Assert.NonNull?

jaredpar commented on src/Test/Utilities/CoreClr/TestExecutionLoadContext.cs:
```diff
@@ -0,0 +1,184 @@
+
+            if (stdOut.Length == 0 && stdErr.Length == 0)
+            {
+            }
```
>} [](start = 12, length = 1)

What's the purpose of this block?

jaredpar commented on src/Test/Utilities/CoreClr/TestExecutionLoadContext.cs:
```diff
@@ -0,0 +1,184 @@
+        internal string DumpAssemblyData()
+        {
+            return "";
+        }
```
Method doesn't appear to be used, can it be deleted?

agocke replied:
Satisfies the interface -- it's supposed to dump the assembly to some output directory if a test fails, but I've never used that feature. I can file a bug.

---
In reply to: [110747253](https://github.com/dotnet/roslyn/pull/18471#discussion_r110747253) [](ancestors = 110747253)

jaredpar replied:
Are you sure?  It's internal hence can't be an interface implementation. 

---
In reply to: [110778296](https://github.com/dotnet/roslyn/pull/18471#discussion_r110778296) [](ancestors = 110778296,110747253)

agocke replied:
Nvm, it's not part of the interface, just called from another method. I'm removing it for now. It can be implemented if we feel it's necessary.

---
In reply to: [110778296](https://github.com/dotnet/roslyn/pull/18471#discussion_r110778296) [](ancestors = 110778296,110747253)

agocke commented:
>             : new EmitOptions().WithDebugInformationFormat(DebugInformationFormat.PortablePdb);

As far as I can tell this is just an internal test helper. Is there any harming in changing the  tests?

---
In reply to: [293053686](https://github.com/dotnet/roslyn/pull/18471#issuecomment-293053686) [](ancestors = 293053686)

---
Refers to: src/Compilers/Core/Portable/Emit/EmitOptions.cs:17 in da5ba75. [](commit_id = da5ba752b4870d77fd8495e79f15c09aa3873838, deletion_comment = False)

tmat commented on src/Compilers/CSharp/Test/Symbol/Compilation/ReferenceManagerTests.cs:
```diff
@@ -2202,7 +2202,7 @@ public void AsymmetricUnification()
         }
 
-        [Fact]
+        [ConditionalFact(typeof(DesktopOnly))]
```
>ConditionalFact(typeof(DesktopOnly)) [](start = 9, length = 36)

Why desktop only?

tmat commented on src/Compilers/CSharp/Test/Symbol/DocumentationComments/DocumentationCommentCompilerTests.cs:
```diff
@@ -2068,7 +2068,7 @@ partial class C
         }
 
-        [ClrOnlyFact(ClrOnlyReason.DocumentationComment)]
+        [ConditionalFact(typeof(ClrOnly), typeof(DesktopOnly))]
```
>typeof(ClrOnly), typeof(DesktopOnly) [](start = 25, length = 36)

Perhaps include a comment, why the test is not xplat.

tmat commented on src/Scripting/Core/ScriptOptions.cs:
```diff
@@ -63,6 +63,11 @@ private static ImmutableArray<MetadataReference> GetDefaultMetadataReferences()
+                "System.Linq.Expressions",
+                "System.Runtime.Numerics",
+                "System.Dynamic.Runtime",
+                "Microsoft.CSharp",
```
Why do we need these?

agocke replied:
This roughly matches the surface area of DLLs in csi.rsp. Notably, csi.rsp has a MS.CSharp reference while this did not, so you couldn't use dynamic.

---
In reply to: [110797808](https://github.com/dotnet/roslyn/pull/18471#discussion_r110797808) [](ancestors = 110797808)

tmat commented on src/Scripting/Core/ScriptOptions.cs:
```diff
@@ -54,6 +53,7 @@ private static ImmutableArray<MetadataReference> GetDefaultMetadataReferences()
                 "System.Reflection.Extensions",
                 "System.Reflection.Primitives",
                 "System.Runtime",
+                "System.Runtime.Extensions",
```
>m.Runtime.Extensions", [](start = 22, length = 22)

Why do we need this?

agocke replied:
Needed for `dynamic`

---
In reply to: [110797820](https://github.com/dotnet/roslyn/pull/18471#discussion_r110797820) [](ancestors = 110797820)

tmat replied:
Are any tests failing without it? Note that scripting references are recursive, so if Microsoft.CSharp.dll references System.Runtime.Extensions it will be automatically referenced.

agocke replied:
My mistake, it's actually not `dynamic`, it's `Environment`. It looks like two scripting tests are failing without it: `CompilationChain_UsingDuplicates` and `CompilationChain_GlobalImports

---
In reply to: [110970347](https://github.com/dotnet/roslyn/pull/18471#discussion_r110970347) [](ancestors = 110970347)

agocke replied:
I'm going to try to change the tests to use APIs that are already available in the surface area.

---
In reply to: [110973145](https://github.com/dotnet/roslyn/pull/18471#discussion_r110973145) [](ancestors = 110973145,110970347)

tmat replied:
Odd. I think we had the tests running on CoreCLR already. Not sure why wouldn't it suddenly not work. Or maybe the tests were added since the last time I run them on Core. Can you try to run them on Core without your change to see if they pass?

---
In reply to: [110977656](https://github.com/dotnet/roslyn/pull/18471#discussion_r110977656) [](ancestors = 110977656,110973145,110970347)

tmat replied:
Ah, never mind. I think I know what the issue is.

---
In reply to: [110982988](https://github.com/dotnet/roslyn/pull/18471#discussion_r110982988) [](ancestors = 110982988,110977656,110973145,110970347)

tmat replied:
Let's add System.Runtime.Extensions here and to the csi.rsp

---
In reply to: [110983277](https://github.com/dotnet/roslyn/pull/18471#discussion_r110983277) [](ancestors = 110983277,110982988,110977656,110973145,110970347)

tmat commented on src/Scripting/Core/ScriptOptions.cs:
```diff
@@ -36,7 +36,6 @@ private static ImmutableArray<MetadataReference> GetDefaultMetadataReferences()
             }
 
-            // Provide similar surface to mscorlib (netstandard 2.0).
             // These references are resolved lazily. Keep in sync with list in core csi.rsp.
```
> // These references are resolved lazily. Keep in sync with list in core csi.rsp. [](start = 11, length = 81)

This list should correspond to csi.rsp.

agocke replied:
Right, that makes sense, but this surface area doesn't really look like mscorlib in netstandard 2.0 at all.

---
In reply to: [110797870](https://github.com/dotnet/roslyn/pull/18471#discussion_r110797870) [](ancestors = 110797870)

tmat replied:
"similar" is the keyword :) 
Let's not change it now since that'd be a product change. We can add more references later.

tmat commented on src/Scripting/CSharpTest/CommandLineRunnerTests.cs:
```diff
@@ -58,6 +58,7 @@ private static IEnumerable<string> GetReferences()
                 yield return "System.Reflection.Extensions";
                 yield return "System.Reflection.Primitives";
                 yield return "System.Runtime";
+                yield return "System.Runtime.Extensions";
```
>"System.Runtime.Extensions"; [](start = 29, length = 28)

Why do we need this?

tmat commented:
:clock1:

jaredpar commented on src/Compilers/CSharp/Test/Symbol/DocumentationComments/DocumentationCommentCompilerTests.cs:
```diff
@@ -6149,6 +6165,8 @@ static void Main()
         /// "--" is not valid within an XML comment.
         /// </summary>
+        // DesktopOnly because these tests reflect over the framework System.Xml
+        // type for error analysis
```
Do we need a bug tracking this?

agocke replied:
@jaredpar Definitely. I'll add one if one doesn't already exist.

agocke commented:
@tmat Updated with your comments.

agocke commented:
test windows_eta_open_prtest please

agocke commented:
@jaredpar Do you know what's up with the ETA RepoUtil failure? I try and run `verify` on my machine (even in the internal repo) and it says everything's fine.

jaredpar commented:
Here is the error

``` txt
Error! Package Microsoft.NETCore.App has different versions:
	pen\build\MSBuildToolset\project.json at 1.1.0
	pen\src\Test\DeployCoreClrTestRuntime\project.json at 2.0.0-beta-001836-00
The versions must be the same or one must be explicitly listed as fixed in RepoData.json
Error! Package NETStandard.Library has different versions:
	pen\src\Test\Utilities\CoreClr\project.json at 2.0.0-beta-25123-01
	pen\src\Tools\Source\CompilerGeneratorTools\Source\BoundTreeGenerator\project.json at 1.6.1
The versions must be the same or one must be explicitly listed as fixed in RepoData.json
Error! Package NETStandard.Library has different versions:
	pen\src\Test\Utilities\CoreClr\project.json at 2.0.0-beta-25123-01
	pen\src\Tools\Source\CompilerGeneratorTools\Source\CSharpErrorFactsGenerator\project.json at 1.6.1
The versions must be the same or one must be explicitly listed as fixed in RepoData.json
Error! Package NETStandard.Library has different versions:
	pen\src\Test\Utilities\CoreClr\project.json at 2.0.0-beta-25123-01
	pen\src\Tools\Source\CompilerGeneratorTools\Source\CSharpSyntaxGenerator\project.json at 1.6.1
The versions must be the same or one must be explicitly listed as fixed in RepoData.json
Error! Package NETStandard.Library has different versions:
	pen\src\Test\Utilities\CoreClr\project.json at 2.0.0-beta-25123-01
	pen\src\Tools\Source\CompilerGeneratorTools\Source\VisualBasicErrorFactsGenerator\project.json at 1.6.1
The versions must be the same or one must be explicitly listed as fixed in RepoData.json
Error! Package NETStandard.Library has different versions:
	pen\src\Test\Utilities\CoreClr\project.json at 2.0.0-beta-25123-01
	pen\src\Tools\Source\CompilerGeneratorTools\Source\VisualBasicSyntaxGenerator\project.json at 1.6.1
The versions must be the same or one must be explicitly listed as fixed in RepoData.json
```

Have you changed any of these?  Your PR is too big for me to check.

agocke commented:
@jaredpar I have, but I also added a new line to the RepoData.json to exclude everything in `src/Tools/Source/CompilerGeneratorTools` and `build\MSBuildToolset`. It seems like those lines somehow aren't getting picked up in the full build.

I also had to fixup the source for the RepoUtil tool itself because of a bug in parsing the exclusion paths. These same errors appeared before I fixed up the tool. Is it possible that RepoUtil is being built with old sources somehow?

jaredpar commented:
> Is it possible that RepoUtil is being built with old sources somehow?

No it's all live. 

Are you using the exact command line that's being used in the internal job?  It's slightly different when run internally.

agocke commented:
@jaredpar Perhaps not. I'll dig in more.

agocke commented:
@jaredpar It looks like the internal repo has a different base root than Open. I've added you on that PR to confirm that these should be the same.

agocke commented:
test mac please

agocke commented:
test mac please

jaredpar commented:
Note that you won't be able to test mac against the new image until the netci.groovy change is commited and the jobs are re-generated.

agocke commented:
@jaredpar Gotcha. I manually edited the job temporarily to force it to 10.12. If this passes I think we're good.

agocke commented:
test windows_eta_open_prtest please
